### PR TITLE
feat(desktop): archive workspaces with a 30-day auto-delete

### PR DESCRIPTION
**Pull request by tao101**
Archive finished worktree workspaces in one click, with Undo and Restore. Archived workspaces have a dedicated sidebar view and board column; opening one shows a Restore banner. Running terminals and agents continue after archiving.

After 30 days, eligible worktrees and their local branches are deleted. Uncommitted, unpushed, or unverifiable work pauses deletion.

## Screenshots

**Archived sidebar and board**

Archived sidebar and board

**Archive with Undo**

Archive confirmation with Undo

**Restore from an archived workspace**

Archived workspace banner with Restore

## Verification

Lint, typecheck, i18n, and plugin checks passed. Host-service and desktop tests have zero new failures against trunk (8 existing GitWatcher and 2 terminal transport failures remain). Live UI checks covered Archive → Undo, Restore, keyboard focus, and the narrow-window banner.

### Current behavior
Archive removes a worktree's files and stops its workspace resources while preserving its branch and archived workspace record. Restore recreates the worktree at its original path and reuses the same workspace identity. Permanent Delete makes that record non-restorable and can delete its branch.

The sidebar and workspace board expose archived workspaces, host selection is retained for lifecycle actions, and sidebar/palette actions support Archive and Undo with warning feedback. Concurrent palette requests for different workspaces are retained. Archived entries are excluded from live navigation shortcuts and grouping actions. Hosts without the dedicated Archive API show an update error before any destructive operation.

This reviewed implementation supersedes the earlier 30-day retention description above: it has no scheduled 30-day purge, and archiving does not leave terminals running. The original title, screenshots, and earlier verification section are historical; they do not establish rendered UI verification for this commit.

### Verification
All six local layers and the combined feature completed Codex review (26 recorded rounds, including 10 union rounds). The ledger records 20 fixed items and four declined findings, with no unresolved review items. The combined feature is reviewed against pinned trunk `277e26129402`. Verified fixes cover archive/revive races, unsupported session archives, worktree ownership, symlink paths, missing locked worktrees, worker execution, warning feedback, and concurrent palette actions.

Final full gate at `7fa9772d441a` passed under an explicit no-new-failures policy: monorepo lint/type checks, strict translation compilation, `check:i18n`, `check:plugins`, and catalog consistency passed. Host-service tests: 1,780 passed, 1 skipped, 9 failed across 169 files run in separate bounded processes. Desktop tests: 3,778 passed, 1 skipped, 2 failed across 414 files. All 50 archive/revival integration tests passed, as did the archive hook and sidebar action tests.

The failures are retained as failures: eight host GitWatcher timer tests and two desktop transport timer tests match the frozen-trunk baseline; one unchanged GitWatcher lazy-registration timeout was independently adjudicated out of scope. Test commands returned nonzero, so this is a qualified gate, not an entirely green suite. Native package test commands were used because the generic full-check adapter forwards incompatible flags. The desktop recap repeats failed names; qualification counts the two actual failures and retains the full raw log. Two test harness/assertion defects caught by the first final gate were repaired before this run. No live Electron/CDP instance or required local environment was available, so the rendered desktop journey was not exercised during this run.

### Reviewer notes
Try Archive → Undo, restoring from the archived view on the owning host, and permanent deletion of an archived workspace. Restoration intentionally refuses a missing branch, a missing locked worktree, an unrelated replacement repository, a mismatched or detached HEAD, or a branch/path already owned by another live workspace. Try an older host as well: Archive must show the host-update error without deleting anything.

## How to review

The branch contains six ordered layers, with fix commits retained in their owning layers. Review each layer as the range from the preceding layer tip (pinned trunk for layer 1) to its listed tip. Each layer was reviewed in isolation, then the combined feature was reviewed again.

| # | commit | type | layer | files | lines |
|---|--------|------|-------|-------|-------|
| 1 | `0c4a3378f6` | feat | 01-tombstone-archive-and-revival | 9 | 1577 |
| 2 | `2b85807569` | feat | 02-archive-host-targeting | 12 | 375 |
| 3 | `b62400c899` | feat | 03-archived-sidebar-and-board | 17 | 336 |
| 4 | `83f96befee` | feat | 04-archive-undo-and-row-controls | 7 | 1170 |
| 5 | `31ad8bde98` | feat | 05-palette-tombstone-archive | 7 | 163 |
| 6 | `7fa9772d44` | chore | 06-tombstone-archive-translations | 17 | 799 |

## What the review found

- Rounds: 16 layer rounds across 6 layers; 10 combined-feature rounds, for 26 recorded rounds total. The union gate converged after two full gate runs and a final confirmation review.
- Findings raised: 24 (5 from the test gate). Fixed: 20 (of which P2, batched after convergence: 3). Declined: 4 (priority 0, scope 2, copy 0, duplicate 0, refuted 2, nitpick 0). Routed: 0.
- Termination: a layer converges on consecutive clean verified rounds (one if it never needed a fix, two otherwise), then the full test gate once, then one confirmation round; the merged whole follows the same rule. Round cap 30; capped layers: none; parked for a person: none.

## Decisions made by the review loop, not by a person (7)

Read these once; any you would have decided differently is a one-line follow-up.

- `lint` — gate failure: lint (bun run lint): **fix** in `fa24d09d8`. Gate repair: both assertion wraps fixed by independent fixer. Quick lint/typecheck passed, focused commit review completed zero findings. Raw fixer report preserved; gate-result repair path permits recording without fixe
- `test:host-service` — gate failure: test:host-service (timeout --kill-after=10 900 taskset -c 0,1,2,3 env CI=1 bun test --max-concurrency=4): **fix** in `fa24d09d8`. Independent fixer added revival-specific worker task/plumbing and corrected session fixture branch; existing worker behavior unchanged. Quick lint/typecheck and focused commit review passed. Report preserved in agent art
- `apps/desktop/src/renderer/routes/_authenticated/_dashboard/components/DashboardSidebar/components/DashboardSidebarWorkspaceItem/components/DashboardSidebarExpandedWorkspaceRow/DashboardSidebarExpandedWorkspaceRow.tsx:458` — Protect live worktrees when deleting an archived workspace: **fix** in `8b4c8311b`. Policy accepts verifier recommended host-contract fix, as explicitly scoped in fixer brief: new archived Delete route exposes pre-existing cleanup to adopted worktrees. Required additions to the already-owned cleanup con
- `packages/host-service/src/trpc/router/workspace-cleanup/workspace-cleanup.ts:288` — Validate the actual Git HEAD before allowing archive: **fix** in `a8612945f`. Completed single archive-identity rewrite with follow-up bbf5c91d3 (L7-005) before acceptance. Policy accepts verifier recommended necessary expansion of existing archive guard/worker adapter within explicitly owned L1 i
- `packages/host-service/src/trpc/router/workspace-cleanup/workspace-cleanup.ts:416` — Compare fully qualified branch refs when validating archives: **fix** in `bbf5c91d3`. Policy accepts independently recommended minimal archive-specific worker integration, explicitly authorized in owning L1 scope. Rewriter reports pre-existing plumbing touched; existing generic refs contract remains uncha
- `test:host-service` — gate failure: test:host-service (timeout --kill-after=10 1800 taskset -c 0,1,2,3 env CI=1 python3 /home/taoufiq-lotfi/.superset/projects/superset/.git/split-and-review/with-archive-and-delete-option/run-host-isolated.py): **fix** in `0c4a3378f`. Independent gate fixer identified four failures as updatedAt refreshed intentionally on tombstone rollback. Test now bounds updatedAt while retaining exact equality of all other fields and archivedAt. No production chang
- `test:desktop` — gate failure: test:desktop (timeout --kill-after=10 900 taskset -c 0,1,2,3 env CI=1 bun test --max-concurrency=4): **fix** in `768ab8fb8`. Independent fixer identified missing useWorkspaceHostUrl export in the new archive hook test mock; supplied export narrows ready state and returns null otherwise, matching runtime contract. Mock completeness prevents upp

## Assumptions

- The layers deploy together as one change; a finding that only matters if a layer shipped alone was not treated as a defect.
- Invariant kept across layers: Archive sends archive=true and deleteBranch=false, restores only archived tombstones at original path and identity; final Delete makes them non-restorable. (layers 1, 2, 4, 5).
- Invariant kept across layers: Tombstones stay outside live workspace data; archive views load explicit archives and carry the owning host through restore/delete. (layers 1, 2, 3, 4).
- Invariant kept across layers: Archive and Undo are coordinated by workspace identity across sidebar/palette; navigation and keyboard focus survive removal and restore. (layers 3, 4, 5).
- Invariant kept across layers: All archive message identifiers and placeholders have translations in every shipping locale. (layers 3, 4, 5, 6).

Original tip of the branch before the review: `11ba230cb30421ac4ef738d15f1f3f1c1cf3803e` (kept as `sarp-oracle-with-archive-and-delete-option-r3`).

### CODE REVIEW

File: apps/desktop/src/renderer/routes/_authenticated/_dashboard/components/DashboardSidebar/components/DashboardSidebarArchivedWorkspaces/components/ArchivedWorkspaceRow/ArchivedWorkspaceRow.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+			toast.error(errorMessage(error));
+			setIsRestoring(false);
+			return;
+		}
+		setIsRestoring(false);
+		navigate({
+			to: "/v2-workspace/$workspaceId",
+			params: { workspaceId: workspace.id },
+		});
+	};
```
[the remark this answers is not in the document]
Comment (tao101) [reply apps/desktop/src/renderer/routes/_authenticated/_dashboard/components/DashboardSidebar/components/DashboardSidebarArchivedWorkspaces/components/ArchivedWorkspaceRow/ArchivedWorkspaceRow.tsx]: Fixed in 1e7046186: the popover is now controlled and closes itself before the restore navigates.

File: apps/desktop/src/renderer/routes/_authenticated/_dashboard/components/DashboardSidebar/components/DashboardSidebarWorkspaceItem/DashboardSidebarWorkspaceItem.tsx [hunk 2]
```diff
@@ -97,8 +97,12 @@ export function DashboardSidebarWorkspaceItem({
 	} = workspace;
 	const isMainWorkspace = workspace.type === "main";
 	const isSessionWorkspace = workspace.type === "session";
+	// Cloud rows have no host row to shelve; the host would refuse anyway.
+	const canArchive = workspace.type === "worktree" && hostType !== "cloud";
```
[the remark this answers is not in the document]
Comment (tao101) [reply apps/desktop/src/renderer/routes/_authenticated/_dashboard/components/DashboardSidebar/components/DashboardSidebarWorkspaceItem/DashboardSidebarWorkspaceItem.tsx]: Fixed in 1e7046186: the context-menu Archive item is withheld when the host is offline; the row keeps its offline tooltip.

File: packages/host-service/src/runtime/shelved-workspace-purge.ts [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+					continue;
+				}
+			}
+			try {
+				await destroy(ctx, {
+					workspaceId: row.id,
+					deleteBranch: false,
+					force: false,
+					teardownMode: "best-effort",
+				});
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/host-service/src/runtime/shelved-workspace-purge.ts]: Fixed in 1e7046186: the sweep re-reads the row after the worktree check and skips destroy unless shelvedAt still matches and archivedAt is null. Regression test added for an unshelve during the check.

File: packages/host-service/src/runtime/shelved-workspace-purge.ts [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+ * Unref'd like the terminal reaper: a pending tick must never hold the
+ * process open, which is also why it needs no shutdown hook.
+ */
+export function startShelvedWorkspacePurge(ctx: HostServiceContext): void {
+	const interval = setInterval(() => {
+		void runShelvedWorkspacePurge(ctx).catch((err) => {
+			console.warn("[shelved-workspace-purge] sweep failed:", err);
+		});
+	}, SHELF_PURGE_INTERVAL_MS);
+	interval.unref();
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/host-service/src/runtime/shelved-workspace-purge.ts]: Fixed in 1e7046186: startShelvedWorkspacePurge returns a stop function and createApp.dispose() calls it first.

File: packages/host-service/src/workspaces/local-workspace-store.ts [hunk 5]
```diff
@@ -361,6 +361,78 @@ export function unarchiveLocalWorkspace(
...
+	if (!existing) return undefined;
+	if (existing.shelvedAt == null) {
+		ctx.db
+			.update(workspaces)
+			.set({ shelvedAt: Date.now(), updatedAt: Date.now() })
+			.where(eq(workspaces.id, id))
+			.run();
+	}
+	const row = getLocalWorkspace(ctx.db, id);
+	if (row) emitWorkspaceChanged(ctx, "updated", row);
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/host-service/src/workspaces/local-workspace-store.ts]: Fixed in 1e7046186: WorkspaceSnapshot carries shelvedAt and purgeBlockedReason; the desktop applies them when present and falls back to the cached values for older hosts.

File: packages/i18n/locales/de/messages.po [hunk 6]
```diff
@@ -13928,6 +13965,19 @@ msgstr "Damit werden die Sitzung und ihr zugrunde liegender Prozess beendet."
...
 
+#. placeholder {0}: formatRelativeTime(deleteAt)
+msgid "This workspace is archived · deletes {0}"
+msgstr "Dieser Arbeitsbereich ist archiviert · wird {0} gelöscht"
+
+msgid "This workspace is archived · deletes soon"
+msgstr "Dieser Arbeitsbereich ist archiviert · wird bald gelöscht"
+
+msgid "This workspace is archived · deletion paused: couldn't verify the worktree"
+msgstr "Dieser Workspace ist archiviert · Löschen pausiert: Worktree konnte nicht geprüft werden"
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/i18n/locales/de/messages.po]: Fixed in 1e7046186: uses Arbeitsbereich.

File: packages/i18n/locales/en/messages.po [hunk 7]
```diff
@@ -1866,9 +1869,24 @@ msgstr "apps/web/src/app/design/components/LayoutSection/LayoutSection.tsx"
 msgid "Archive"
 msgstr "Archive"
 
+msgid "Archive · deletes in {SHELF_RETENTION_DAYS, plural, one {# day} other {# days}} unless reopened"
+msgstr "Archive · deletes in {SHELF_RETENTION_DAYS, plural, one {# day} other {# days}} unless reopened"
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/i18n/locales/en/messages.po]: Fixed in 1e7046186: the tooltip now says "unless restored" and every locale was retranslated accordingly.

File: packages/i18n/locales/es/messages.po [hunk 8]
```diff
@@ -13928,6 +13965,19 @@ msgstr "Esto terminará la sesión y su proceso subyacente."
...
 
+#. placeholder {0}: formatRelativeTime(deleteAt)
+msgid "This workspace is archived · deletes {0}"
+msgstr "Este espacio de trabajo está archivado · se elimina {0}"
+
+msgid "This workspace is archived · deletes soon"
+msgstr "Este espacio de trabajo está archivado · se elimina pronto"
+
+msgid "This workspace is archived · deletion paused: couldn't verify the worktree"
+msgstr "Este workspace está archivado · eliminación en pausa: no se pudo verificar el worktree"
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/i18n/locales/es/messages.po]: Fixed in 1e7046186: espacio de trabajo.

File: packages/i18n/locales/fr/messages.po [hunk 9]
```diff
@@ -1866,9 +1869,24 @@ msgstr "apps/web/src/app/design/components/LayoutSection/LayoutSection.tsx"
 msgid "Archive"
 msgstr "Archiver"
 
+msgid "Archive · deletes in {SHELF_RETENTION_DAYS, plural, one {# day} other {# days}} unless reopened"
+msgstr "Archiver · supprimé dans {SHELF_RETENTION_DAYS, plural, one {# jour} other {# jours}} sauf réouverture"
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/i18n/locales/fr/messages.po]: Fixed in 1e7046186: the French countdown strings now use "Suppression …" / "Suppression prochaine".

File: packages/i18n/locales/fr/messages.po [hunk 10]
```diff
@@ -13928,6 +13965,19 @@ msgstr "Cette action arrêtera la session et son processus sous-jacent."
...
 
+#. placeholder {0}: formatRelativeTime(deleteAt)
+msgid "This workspace is archived · deletes {0}"
+msgstr "Cet espace de travail est archivé · supprimé {0}"
+
+msgid "This workspace is archived · deletes soon"
+msgstr "Cet espace de travail est archivé · supprimé bientôt"
+
+msgid "This workspace is archived · deletion paused: couldn't verify the worktree"
+msgstr "Ce workspace est archivé · suppression en pause : impossible de vérifier le worktree"
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/i18n/locales/fr/messages.po]: Fixed in 1e7046186: Cet espace de travail.

File: packages/i18n/locales/nl/messages.po [hunk 11]
```diff
@@ -13928,6 +13965,19 @@ msgstr "Hiermee worden de sessie en het onderliggende proces beëindigd."
...
 
+#. placeholder {0}: formatRelativeTime(deleteAt)
+msgid "This workspace is archived · deletes {0}"
+msgstr "Deze werkruimte is gearchiveerd · wordt {0} verwijderd"
+
+msgid "This workspace is archived · deletes soon"
+msgstr "Deze werkruimte is gearchiveerd · wordt binnenkort verwijderd"
+
+msgid "This workspace is archived · deletion paused: couldn't verify the worktree"
+msgstr "Deze workspace is gearchiveerd · verwijderen gepauzeerd: worktree kon niet worden gecontroleerd"
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/i18n/locales/nl/messages.po]: Fixed in 1e7046186: werkruimte.

File: packages/i18n/locales/pl/messages.po [hunk 12]
```diff
@@ -13928,6 +13965,19 @@ msgstr "Zakończy to sesję i jej proces bazowy."
...
 
+#. placeholder {0}: formatRelativeTime(deleteAt)
+msgid "This workspace is archived · deletes {0}"
+msgstr "Ten obszar roboczy jest zarchiwizowany · zostanie usunięty {0}"
+
+msgid "This workspace is archived · deletes soon"
+msgstr "Ten obszar roboczy jest zarchiwizowany · wkrótce zostanie usunięty"
+
+msgid "This workspace is archived · deletion paused: couldn't verify the worktree"
+msgstr "Ten workspace jest zarchiwizowany · usuwanie wstrzymane: nie udało się zweryfikować worktree"
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/i18n/locales/pl/messages.po]: Fixed in 1e7046186: obszar roboczy.

File: packages/i18n/locales/ru/messages.po [hunk 13]
```diff
@@ -13928,6 +13965,19 @@ msgstr "Это завершит сессию и связанный с ней п
...
 
+#. placeholder {0}: formatRelativeTime(deleteAt)
+msgid "This workspace is archived · deletes {0}"
+msgstr "Эта рабочая область архивирована · будет удалена {0}"
+
+msgid "This workspace is archived · deletes soon"
+msgstr "Эта рабочая область архивирована · скоро будет удалена"
+
+msgid "This workspace is archived · deletion paused: couldn't verify the worktree"
+msgstr "Этот workspace в архиве · удаление приостановлено: не удалось проверить worktree"
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/i18n/locales/ru/messages.po]: Fixed in 1e7046186: рабочая область.

File: packages/i18n/locales/tr/messages.po [hunk 14]
```diff
@@ -13928,6 +13965,19 @@ msgstr "Bu işlem oturumu ve altındaki süreci sonlandırır."
...
 
+#. placeholder {0}: formatRelativeTime(deleteAt)
+msgid "This workspace is archived · deletes {0}"
+msgstr "Bu çalışma alanı arşivlendi · {0} silinecek"
+
+msgid "This workspace is archived · deletes soon"
+msgstr "Bu çalışma alanı arşivlendi · yakında silinecek"
+
+msgid "This workspace is archived · deletion paused: couldn't verify the worktree"
+msgstr "Bu workspace arşivlendi · silme duraklatıldı: worktree doğrulanamadı"
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/i18n/locales/tr/messages.po]: Fixed in 1e7046186: çalışma alanı.

File: packages/i18n/locales/zh-CN/messages.po [hunk 15]
```diff
@@ -4498,6 +4519,13 @@ msgstr "已删除或不可用，请另选一个"
 msgid "Deleted workspaces, but couldn’t finish updating the sidebar"
 msgstr "工作区已删除，但侧边栏没能更新完"
 
+#. placeholder {0}: formatRelativeTime(workspace.deleteAt)
+msgid "Deletes {0}"
+msgstr "{0}删除"
```
[the remark this answers is not in the document]
Comment (tao101) [reply packages/i18n/locales/zh-CN/messages.po]: Not applying: this countdown uses formatRelativeTime from @superset/i18n/format, which is Intl.RelativeTimeFormat with numeric: auto, so in zh-CN it renders e.g. 12天后删除 / 明天删除, not English tokens. The compact desktop helper you found is a different function that this code does not import.

**Comment by tao101** [comment]
@Kitenite Went with your suggestion: no separate shelved state. Archive is the existing destroy with the branch kept, stamped `archiveReason: "archived"`. Restore is a new `workspaceCleanup.revive` that checks the branch out again at the tombstone's original path and clears `archivedAt`, so chats keyed by that path come back. No new columns, no migration, no 30-day purge.

- Sidebar Archived view and board Archived lane show archived tombstones only; plain deletes stay in Deleted.
- Archived rows offer Restore and Delete. Delete uses the normal dialog with the branch checkbox and makes the tombstone final.
- Revive refuses a missing branch, a branch checked out elsewhere, or an occupied path.

Host integration tests cover archive, revive, delete-of-archive and the refusals. Not yet verified in the running app.
